crates/sui-core/src/consensus_throughput_calculator.rs

```diff
@@ -0,0 +1,663 @@
+    /// Holds the throughput profiles by the throughput range (upper_throughput, cool_down_threshold)
+    profiles: BTreeMap<u64, ThroughputProfile>,
+}
+
+impl ThroughputProfileRanges {
+    pub fn new(profiles: &[ThroughputProfile]) -> Self {
+        let mut p: BTreeMap<u64, ThroughputProfile> = BTreeMap::new();
+
+        for profile in profiles {
+            assert!(
+                !p.iter().any(|(_, pr)| pr.level == profile.level),
+                "Attempted to insert profile with same level"
+            );
+            assert!(
+                p.insert(profile.throughput, *profile).is_none(),
+                "Attempted to insert profile with same throughput"
+            );
+        }
+
+        // By default the Low profile should exist with throughput 0
+        assert_eq!(
+            *p.get(&0).unwrap(),
+            ThroughputProfile {
+                level: Level::Low,
+                throughput: 0
+            }
+        );
+
+        Self { profiles: p }
+    }
+
+    pub fn lowest_profile(&self) -> ThroughputProfile {
+        *self
+            .profiles
+            .first_key_value()
+            .expect("Should contain at least one throughput profile")
+            .1
+    }
+
+    pub fn highest_profile(&self) -> ThroughputProfile {
+        *self
+            .profiles
+            .last_key_value()
+            .expect("Should contain at least one throughput profile")
+            .1
+    }
+    /// Resolves the throughput profile that corresponds to the provided throughput.
+    pub fn resolve(&self, current_throughput: u64) -> ThroughputProfile {
+        let mut iter = self.profiles.iter();
+        while let Some((threshold, profile)) = iter.next_back() {
+            if current_throughput >= *threshold {
+                return *profile;
+            }
+        }
+
+        warn!("Could not resolve throughput profile for throughput {} - we shouldn't end up here. Fallback to lowest profile as default.", current_throughput);
+
+        // If not found, then we should return the lowest possible profile as default to stay on safe side.
+        self.highest_profile()
+    }
+}
+
+impl Default for ThroughputProfileRanges {
+    fn default() -> Self {
+        let profiles = vec![
+            ThroughputProfile {
+                level: Level::Low,
+                throughput: 0,
+            },
+            ThroughputProfile {
+                level: Level::High,
+                throughput: 2_000,
+            },
+        ];
+        ThroughputProfileRanges::new(&profiles)
+    }
+}
+
+pub type TimestampSecs = u64;
+
+#[derive(Debug, Copy, Clone)]
+pub struct ThroughputProfileEntry {
+    /// The throughput profile
+    profile: ThroughputProfile,
+    /// The time when this throughput profile was created
+    timestamp: TimestampSecs,
+    /// The calculated throughput when this profile created
+    throughput: u64,
+}
+
+#[derive(Default)]
+struct ConsensusThroughputCalculatorInner {
```

Comment:
IIUC this is basically holding the rolling window so can we name it accordingly and then provide the impl to maintain that window? I was thinking something like this

```
struct RollingWindow {
    observations: Vec<(TimestampSecs, u64)>,
    window_size: u64,
}

impl RollingWindow {
    fn new(window_size: u64) -> Self {
        Self {
            observations: Vec::new(),
            window_size,
        }
    }

    fn add_observation(&mut self, timestamp_secs: TimestampSecs, num_of_transactions: u64) {
        self.observations.push((timestamp_secs, num_of_transactions));
        if self.observations.len() > self.window_size {
            self.observations.remove(0);
        }
    }

    fn calculate_throughput(&self) -> Option<u64> {
        if self.observations.len() == 0 {
            return None;
        }

        let first_element_ts = self.observations[0].0;
        let last_element_ts = self.observations[self.observations.len() - 1].0;

        let period = first_element_ts.saturating_sub(last_element_ts);

        Some(self.observations.iter().fold(0, |acc, (timestamp_secs, num_of_transactions)| {
            acc + num_of_transactions / period
        }))
    }
}
```

I think that will make add_transactions() a little simpler. 

```
pub fn add_transactions(&self, timestamp_ms: TimestampMs, num_of_transactions: u64) {
        let timestamp_secs = timestamp_ms / 1_000;

        self.rolling_window.add_observation(timestamp_secs, num_of_transactions);

        if let Some(throughput) = self.rolling_window.calculate_throughput() {
            self.metrics.consensus_calculated_throughput.set(throughput as i64);
        }
    }
```

I am missing some of the time checks and other protections you have but thats the rough idea